Panic when publishing to a path with runOnDemand without any clients reading

<!--
To increase the chance of your issue getting fixed, open an issue FOR EACH problem. Please do not report multiple problems in a single issue, otherwise they'll probably never get ALL fixed.

Please include all sections of this template into your issue, or it will be automatically closed.
-->

## Which version are you using?

v1.2.1

## Which operating system are you using?

<!-- fill checkboxes with a x. Example: [x] Linux -->

- [x] Linux amd64 standard
- [x] Linux amd64 Docker
- [x] Linux arm64 standard
- [ ] Linux arm64 Docker
- [ ] Linux arm7 standard
- [ ] Linux arm7 Docker
- [ ] Linux arm6 standard
- [ ] Linux arm6 Docker
- [ ] Windows amd64 standard
- [ ] Windows amd64 Docker (WSL backend)
- [x] macOS amd64 standard
- [ ] macOS amd64 Docker
- [ ] Other (please describe)

## Describe the issue

MTX panics if `runOnDemand` is set for a path and publish to the path before a client starts reading from that path.

This was not an issue in v1.2.0 which would correctly terminate the stream.

## Describe how to replicate the issue

<!--
the maintainers must be able to REPLICATE your issue to solve it - therefore, describe in a very detailed way how to replicate it.
-->

1. set `runOnDemand` for a specific path
2. start the server
3. publish to path before the client starts reading

or 
```
MTX_PATHS_MYPATH_RUNONDEMAND='echo hi' ./mediamtx   
```

## Did you attach the server logs?

<!--
Server logs are sometimes useful to identify the issue.
If you think this is the case, set the parameter 'logLevel' to 'debug' and attach the server logs.
-->

yes 

```
INF [RTSP] [conn 127.0.0.1:54971] opened
INF [RTSP] [session 3487b1c0] created by 127.0.0.1:54971
INF [RTSP] [session 3487b1c0] is publishing to path 'mypath', 1 track (H264)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x187c8a3]

goroutine 7 [running]:
github.com/bluenviron/mediamtx/internal/core.(*path).onDemandPublisherStop(0xc00009a1e0, {0x1ac6138, 0x14})
	/s/internal/core/path.go:869 +0x143
github.com/bluenviron/mediamtx/internal/core.(*path).doOnDemandPublisherCloseTimer(...)
	/s/internal/core/path.go:504
github.com/bluenviron/mediamtx/internal/core.(*path).runInner(0xc00009a1e0)
	/s/internal/core/path.go:405 +0x653
github.com/bluenviron/mediamtx/internal/core.(*path).run(0xc00009a1e0)
	/s/internal/core/path.go:334 +0x325
created by github.com/bluenviron/mediamtx/internal/core.newPath in goroutine 1
	/s/internal/core/path.go:284 +0x676
```

## Did you attach a network dump?

<!--
If the bug arises when using MediaMTX with an external hardware or software, the most helpful content you can provide is a dump of the data exchanged between the server and the target (network dump), that can be generated in this way:
1) Download wireshark (https://www.wireshark.org/)
2) Start capturing on the interface used for exchanging RTSP (if the server and the target software are both installed on your pc, the interface is probably "loopback", otherwise it's the one of your network card)
3) Start the server and replicate the issue
4) Stop capturing, save the result in .pcap format
5) Attach
-->

no

--- a/README.md
+++ b/README.md
@@ -1272,6 +1272,7 @@ The server allows to specify commands that are executed when a certain event hap
 `runOnConnect` allows to run a command when a client connects to the server:
 
 ```yml
+# Command to run when a client connects to the server.
 # This is terminated with SIGINT when a client disconnects from the server.
 # The following environment variables are available:
 # * RTSP_PORT: RTSP server port
@@ -1285,6 +1286,7 @@ runOnConnectRestart: no
 `runOnDisconnect` allows to run a command when a client disconnects from the server:
 
 ```yml
+# Command to run when a client disconnects from the server.
 # Environment variables are the same of runOnConnect.
 runOnDisconnect: curl http://my-custom-server/webhook?conn_type=$MTX_CONN_TYPE&conn_id=$MTX_CONN_ID
 ```
@@ -1294,7 +1296,8 @@ runOnDisconnect: curl http://my-custom-server/webhook?conn_type=$MTX_CONN_TYPE&c
 ```yml
 paths:
   mypath:
-    # This is terminated with SIGINT when the program closes.
+    # Command to run when this path is initialized.
+    # This can be used to publish a stream when the server is launched.
     # The following environment variables are available:
     # * MTX_PATH: path name
     # * RTSP_PORT: RTSP server port
@@ -1310,6 +1313,8 @@ paths:
 ```yml
 paths:
   mypath:
+    # Command to run when this path is requested by a reader
+    # and no one is publishing to this path yet.
     # This is terminated with SIGINT when the program closes.
     # The following environment variables are available:
     # * MTX_PATH: path name
@@ -1326,6 +1331,8 @@ paths:
 
 ```yml
 pathDefaults:
+  # Command to run when the stream is ready to be read, whenever it is
+  # published by a client or pulled from a server / camera.
   # This is terminated with SIGINT when the stream is not ready anymore.
   # The following environment variables are available:
   # * MTX_PATH: path name
@@ -1344,6 +1351,7 @@ pathDefaults:
 
 ```yml
 pathDefaults:
+  # Command to run when the stream is not available anymore.
   # Environment variables are the same of runOnReady.
   runOnNotReady: curl http://my-custom-server/webhook?path=$MTX_PATH&source_type=$MTX_SOURCE_TYPE&source_id=$MTX_SOURCE_ID
 ```
@@ -1352,6 +1360,7 @@ pathDefaults:
 
 ```yml
 pathDefaults:
+  # Command to run when a client starts reading.
   # This is terminated with SIGINT when a client stops reading.
   # The following environment variables are available:
   # * MTX_PATH: path name

--- a/internal/core/path.go
+++ b/internal/core/path.go
@@ -201,8 +201,8 @@ type path struct {
 	chStaticSourceSetReady    chan defs.PathSourceStaticSetReadyReq
 	chStaticSourceSetNotReady chan defs.PathSourceStaticSetNotReadyReq
 	chDescribe                chan pathDescribeReq
-	chRemovePublisher         chan pathRemovePublisherReq
 	chAddPublisher            chan pathAddPublisherReq
+	chRemovePublisher         chan pathRemovePublisherReq
 	chStartPublisher          chan pathStartPublisherReq
 	chStopPublisher           chan pathStopPublisherReq
 	chAddReader               chan pathAddReaderReq
@@ -254,8 +254,8 @@ func newPath(
 		chStaticSourceSetReady:         make(chan defs.PathSourceStaticSetReadyReq),
 		chStaticSourceSetNotReady:      make(chan defs.PathSourceStaticSetNotReadyReq),
 		chDescribe:                     make(chan pathDescribeReq),
-		chRemovePublisher:              make(chan pathRemovePublisherReq),
 		chAddPublisher:                 make(chan pathAddPublisherReq),
+		chRemovePublisher:              make(chan pathRemovePublisherReq),
 		chStartPublisher:               make(chan pathStartPublisherReq),
 		chStopPublisher:                make(chan pathStopPublisherReq),
 		chAddReader:                    make(chan pathAddReaderReq),
@@ -357,7 +357,7 @@ func (pa *path) run() {
 	}
 
 	if pa.onUnDemandHook != nil {
-		pa.onUnDemandHook("path closed")
+		pa.onUnDemandHook("path destroyed")
 	}
 
 	pa.Log(logger.Debug, "destroyed: %v", err)
@@ -414,15 +414,15 @@ func (pa *path) runInner() error {
 				return fmt.Errorf("not in use")
 			}
 
+		case req := <-pa.chAddPublisher:
+			pa.doAddPublisher(req)
+
 		case req := <-pa.chRemovePublisher:
 			pa.doRemovePublisher(req)
 
 			if pa.shouldClose() {
 				return fmt.Errorf("not in use")
 			}
-
-		case req := <-pa.chAddPublisher:
-			pa.doAddPublisher(req)
 
 		case req := <-pa.chStartPublisher:
 			pa.doStartPublisher(req)
@@ -519,21 +519,10 @@ func (pa *path) doSourceStaticSetReady(req defs.PathSourceStaticSetReadyReq) {
 	if pa.conf.HasOnDemandStaticSource() {
 		pa.onDemandStaticSourceReadyTimer.Stop()
 		pa.onDemandStaticSourceReadyTimer = newEmptyTimer()
-
 		pa.onDemandStaticSourceScheduleClose()
-
-		for _, req := range pa.describeRequestsOnHold {
-			req.res <- pathDescribeRes{
-				stream: pa.stream,
-			}
-		}
-		pa.describeRequestsOnHold = nil
-
-		for _, req := range pa.readerAddRequestsOnHold {
-			pa.addReaderPost(req)
-		}
-		pa.readerAddRequestsOnHold = nil
 	}
+
+	pa.consumeOnHoldRequests()
 
 	req.Res <- defs.PathSourceStaticSetReadyRes{Stream: pa.stream}
 }
@@ -649,24 +638,13 @@ func (pa *path) doStartPublisher(req pathStartPublisherReq) {
 		pa.name,
 		mediaInfo(req.desc.Medias))
 
-	if pa.conf.HasOnDemandPublisher() {
+	if pa.conf.HasOnDemandPublisher() && pa.onDemandPublisherState != pathOnDemandStateInitial {
 		pa.onDemandPublisherReadyTimer.Stop()
 		pa.onDemandPublisherReadyTimer = newEmptyTimer()
-
 		pa.onDemandPublisherScheduleClose()
-
-		for _, req := range pa.describeRequestsOnHold {
-			req.res <- pathDescribeRes{
-				stream: pa.stream,
-			}
-		}
-		pa.describeRequestsOnHold = nil
-
-		for _, req := range pa.readerAddRequestsOnHold {
-			pa.addReaderPost(req)
-		}
-		pa.readerAddRequestsOnHold = nil
 	}
+
+	pa.consumeOnHoldRequests()
 
 	req.res <- pathStartPublisherRes{stream: pa.stream}
 }
@@ -840,20 +818,15 @@ func (pa *path) onDemandPublisherScheduleClose() {
 }
 
 func (pa *path) onDemandPublisherStop(reason string) {
-	if pa.source != nil {
-		pa.source.(publisher).close()
-		pa.executeRemovePublisher()
-	}
-
 	if pa.onDemandPublisherState == pathOnDemandStateClosing {
 		pa.onDemandPublisherCloseTimer.Stop()
 		pa.onDemandPublisherCloseTimer = newEmptyTimer()
 	}
 
-	pa.onDemandPublisherState = pathOnDemandStateInitial
-
 	pa.onUnDemandHook(reason)
 	pa.onUnDemandHook = nil
+
+	pa.onDemandPublisherState = pathOnDemandStateInitial
 }
 
 func (pa *path) setReady(desc *description.Session, allocateEncoder bool) error {
@@ -879,6 +852,20 @@ func (pa *path) setReady(desc *description.Session, allocateEncoder bool) error 
 	pa.parent.pathReady(pa)
 
 	return nil
+}
+
+func (pa *path) consumeOnHoldRequests() {
+	for _, req := range pa.describeRequestsOnHold {
+		req.res <- pathDescribeRes{
+			stream: pa.stream,
+		}
+	}
+	pa.describeRequestsOnHold = nil
+
+	for _, req := range pa.readerAddRequestsOnHold {
+		pa.addReaderPost(req)
+	}
+	pa.readerAddRequestsOnHold = nil
 }
 
 func (pa *path) setNotReady() {
@@ -1048,16 +1035,6 @@ func (pa *path) describe(req pathDescribeReq) pathDescribeRes {
 	}
 }
 
-// removePublisher is called by a publisher.
-func (pa *path) removePublisher(req pathRemovePublisherReq) {
-	req.res = make(chan struct{})
-	select {
-	case pa.chRemovePublisher <- req:
-		<-req.res
-	case <-pa.ctx.Done():
-	}
-}
-
 // addPublisher is called by a publisher through pathManager.
 func (pa *path) addPublisher(req pathAddPublisherReq) pathAddPublisherRes {
 	select {
@@ -1065,6 +1042,16 @@ func (pa *path) addPublisher(req pathAddPublisherReq) pathAddPublisherRes {
 		return <-req.res
 	case <-pa.ctx.Done():
 		return pathAddPublisherRes{err: fmt.Errorf("terminated")}
+	}
+}
+
+// removePublisher is called by a publisher.
+func (pa *path) removePublisher(req pathRemovePublisherReq) {
+	req.res = make(chan struct{})
+	select {
+	case pa.chRemovePublisher <- req:
+		<-req.res
+	case <-pa.ctx.Done():
 	}
 }
 

--- a/mediamtx.yml
+++ b/mediamtx.yml
@@ -468,7 +468,8 @@ pathDefaults:
   # Restart the command if it exits.
   runOnInitRestart: no
 
-  # Command to run when this path is requested by a reader.
+  # Command to run when this path is requested by a reader
+  # and no one is publishing to this path yet.
   # This can be used to publish a stream on demand.
   # This is terminated with SIGINT when the path is not requested anymore.
   # The following environment variables are available:

